test/Fencer/Rules/Test.hs

```diff
@@ -227,3 +230,57 @@ test_rulesLoadRulesReadPermissions =
   file2 = simpleRuleFile
     ("domain2" </> "config" </> "config.yml")
     domainDescriptorKeyText
+
+-- | test that faulty rules in reloading with 'loadRulesFromDirectory'
+-- are rejected and the valid ones that were previously loaded are
+-- kept instead.
+--
+-- This matches the behavior of @lyft/ratelimit@.
+test_rulesReloadRules :: TestTree
+test_rulesReloadRules =
+  testCase "Rules fail to reload for an invalid domain" $
+    Temp.withSystemTempDirectory "fencer-config" $ \tempDir -> do
+      rules <- writeAndLoadRules
+        (#ignoreDotFiles False)
+        (#root tempDir)
+        (#files files)
+      failures <- writeAndLoadRules
+        (#ignoreDotFiles False)
+        (#root tempDir)
+        (#files files')
+      case (rules, failures) of
+        (Left errs, _) ->
+          assertFailure
+            ("Expected domains, got failures: " ++
+             prettyPrintErrors (NE.toList errs))
+        (Right _, Right newDefinitions) ->
+          assertFailure
+            ("Expected failures, got domains: " ++ show newDefinitions)
+        (Right definitions, f@(Left _)) -> do
+          assertEqual
+            "unexpected failure"
+            (length . toErrorList $ mapLeft NE.toList f)
+            (length . toErrorList $ expectedFailure)
+          assertBool "unexpected definitions"
+            (((==) `on` show)
+             (sortOn domainDefinitionId <$> result)
+             (Right $ sortOn domainDefinitionId definitions))
```

Comment:
It's a little hard to read these lines. I'd probably use do-notation, but even just

```haskell
          assertBool "unexpected definitions" $
            ((==) `on` show)
              (sortOn domainDefinitionId <$> result)
              (Right $ sortOn domainDefinitionId definitions)
```

is already more readable.

Reply:
That's a fair point! I addressed it with `let`s in an upcoming commit.

Reply:
@effectfully: Any remaining issues to be addressed?